### PR TITLE
Make SSL_SESSION safer to transfer between systems with different sizes for |long|

### PR DESCRIPTION
**Pull request by levitte**
This implements internal ASN.1 types INT32, UINT32, INT64, UINT64 and Z variants of them, and replaces alls uses of `long` in `SSL_SESSION_ASN1` with integer types with precise sizes.

This can be seen as an alternative to #3042 that solves the deeper issue.

### CODE REVIEW

**Comment by levitte** [comment]
This is made so that it can be backported as is to 1.1.0.  It *does* add a few more symbols to `util/libcrypto.num`, but those functions are otherwise kept strictly private.  Another PR can make them public for master only.

File: ssl/ssl_asn1.c [hunk 1]
```diff
@@ -37,25 +37,25 @@
...
+    int64_t timeout;
     X509 *peer;
     ASN1_OCTET_STRING *session_id_context;
-    long verify_result;
+    int32_t verify_result;
     ASN1_OCTET_STRING *tlsext_hostname;
-    long tlsext_tick_lifetime_hint;
-    long tlsext_tick_age_add;
+    int64_t tlsext_tick_lifetime_hint;
+    int64_t tlsext_tick_age_add;
```
Comment (mattcaswell) [inline ssl/ssl_asn1.c]: uint32_t

File: ssl/ssl_asn1.c [hunk 2]
```diff
@@ -37,25 +37,25 @@
...
+    int64_t time;
+    int64_t timeout;
     X509 *peer;
     ASN1_OCTET_STRING *session_id_context;
-    long verify_result;
+    int32_t verify_result;
     ASN1_OCTET_STRING *tlsext_hostname;
-    long tlsext_tick_lifetime_hint;
-    long tlsext_tick_age_add;
+    int64_t tlsext_tick_lifetime_hint;
```
Comment (mattcaswell) [inline ssl/ssl_asn1.c]: uint64_t

File: ssl/ssl_asn1.c [hunk 3]
```diff
@@ -37,25 +37,25 @@
...
 #include <stdlib.h>
 #include "ssl_locl.h"
-#include <openssl/asn1t.h>
+#include "internal/asn1t.h"
 #include <openssl/x509.h>
 
 typedef struct {
-    long version;
-    long ssl_version;
+    int32_t version;
```
Comment (mattcaswell) [inline ssl/ssl_asn1.c]: unsigned I'd say

File: ssl/ssl_asn1.c [hunk 4]
```diff
@@ -64,37 +64,37 @@ typedef struct {
...
 #ifndef OPENSSL_NO_SRP
     ASN1_EXP_OPT(SSL_SESSION_ASN1, srp_username, ASN1_OCTET_STRING, 12),
 #endif
-    ASN1_EXP_OPT(SSL_SESSION_ASN1, flags, ZLONG, 13),
-    ASN1_EXP_OPT(SSL_SESSION_ASN1, tlsext_tick_age_add, ZLONG, 14),
-    ASN1_EXP_OPT(SSL_SESSION_ASN1, max_early_data, ZLONG, 15),
+    ASN1_EXP_OPT(SSL_SESSION_ASN1, flags, ZUINT64, 13),
+    ASN1_EXP_OPT(SSL_SESSION_ASN1, tlsext_tick_age_add, ZINT64, 14),
+    ASN1_EXP_OPT(SSL_SESSION_ASN1, max_early_data, ZUINT32, 15),
     ASN1_EXP_OPT(SSL_SESSION_ASN1, alpn_selected, ASN1_OCTET_STRING, 16)
```
Comment (mattcaswell) [inline ssl/ssl_asn1.c]: Equivalent changes here as per my suggestions above.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Looks good. This needs tests though. Feel free to take the one from #3042 which adds tests for LONG/ZLONG and extend it for the new types. Or start from scratch if you prefer.

File: ssl/ssl_asn1.c (hunk 3 above)
Comment (levitte) [inline ssl/ssl_asn1.c]: I think I tried that, and got warnings all over the place because of signed / unsigned comparisons.  I think that adjustment should be left to another PR.

Comment (mattcaswell) [inline ssl/ssl_asn1.c]: For version? I could understand that for ssl_version. Less so for version which I don't think is referred to much. But ok.

Comment (levitte) [inline ssl/ssl_asn1.c]: I tried again, and it works like a charm...  Maybe I confused it with something else...

**Comment by levitte** [comment]
Fixed up all the types you asked about.  Now looking at tests.  I think I'll start over, but thanks for the inspiration!

**Comment by levitte** [comment]
Speaking of which, I'm not sure I understand these lines, mind explaining them?

``` C
unsigned char test_long_prefix[] = { 0x02 };
unsigned char test_long_suffix[] = { 0x02, 0x01, 0x01 };
unsigned char test_zlong_prefix[] = { 0x02, 0x01, 0x00, 0x02 };
```

**Comment by levitte** [comment]
(never mind my previous comment (now removed)...  I got it)

**Comment by levitte** [comment]
Test added and a few bugs uncovered (including in x_long.c)

Let's see what the builds say.

**Comment by levitte** [comment]
Ok, all builds seem to pass now

**Comment by levitte** [comment]
Ping?

File: test/asn1_encode_test.c [hunk 5]
```diff
@@ -0,0 +1,641 @@
...
+    CUSTOM_EXPECTED_SUCCESS(LONG_MAX, LONG_MAX), /* t_8bytes_2 */
+    CUSTOM_EXPECTED_FAILURE,     /* t_8bytes_3_pad (illegal padding) */
+    CUSTOM_EXPECTED_SUCCESS(LONG_MIN, LONG_MIN), /* t_8bytes_4_neg */
+    CUSTOM_EXPECTED_FAILURE,     /* t_8bytes_5_negpad (illegal padding) */
+    CUSTOM_EXPECTED_SUCCESS(0x1ffffffff, 0x1ffffffff), /* t_5bytes_1 */
+    /*
+     * Note: because t_4bytes_1 doesn't have a negative indication, we need
+     * to code the expectation with explicit constants which are negative on
+     * 32-bit platforms and positive on larger ones.
+     */
```
Comment (mattcaswell) [inline test/asn1_encode_test.c]: I can't quite correlate this comment with the equivalent entry for the not SIXTY_FOUR_BIT_LONG case below.

File: test/asn1_encode_test.c [hunk 6]
```diff
@@ -0,0 +1,641 @@
...
+    d2i_fn *d2i;
+    /* Function to free a decoded structure */
+    ifree_fn *ifree;
+} TEST_PACKAGE;
+
+/* To facilitate the creation of an encdec_data array */
+#define ENCDEC_DATA(num, znum)                  \
+    { 0xff, num, 1 }, { 0xff, 1, znum }
+#define ENCDEC_ARRAY(max, zmax)                 \
+    ENCDEC_DATA(max,zmax),                      \
```
Comment (mattcaswell) [inline test/asn1_encode_test.c]: Why "max" values here - but no "min" values?

File: test/asn1_encode_test.c [hunk 7]
```diff
@@ -0,0 +1,641 @@
...
+    CUSTOM_EXPECTED_FAILURE,     /* t_5bytes_1 */
+    CUSTOM_EXPECTED_FAILURE,     /* t_4bytes_1 (too large positive) */
+#endif
+    CUSTOM_EXPECTED_SUCCESS(INT32_MAX - 1, INT32_MAX -1), /* t_4bytes_2 */
+    CUSTOM_EXPECTED_FAILURE,     /* t_4bytes_3_pad (illegal padding) */
+    CUSTOM_EXPECTED_SUCCESS(INT32_MIN, INT32_MIN), /* t_4bytes_4_neg */
+    CUSTOM_EXPECTED_FAILURE,     /* t_4bytes_5_negpad (illegal padding) */
+};
+static ASN1_LONG_DATA long_encdec_data[] = {
+    ENCDEC_ARRAY(LONG_MAX - 1, LONG_MAX),
```
Comment (mattcaswell) [inline test/asn1_encode_test.c]: On systems with a 64-bit int, the max LONG value is LONG_MAX not "LONG_MAX - 1"...which is kind of weird now that I think about it, because ASN1_LONG_UNDEF != LONG_MAX. Should we just define ASN1_LONG_UNDEF to be LONG_MAX? But that could cause 32-bit/64-bit interoperability problems.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/asn1_encode_test.c [hunk 8]
```diff
@@ -0,0 +1,641 @@
...
+#include "internal/asn1t.h"
+#include "internal/numbers.h"
+#include "test_main.h"
+#include "testutil.h"
+
+#ifdef __GNUC__
+# pragma GCC diagnostic ignored "-Wunused-function"
+# pragma GCC diagnostic ignored "-Wformat"
+#endif
+#ifdef __clang_
```
Comment (dot-asm) [inline test/asn1_encode_test.c]: It's `__clang__`, not `__clang_`. But it works out, because clang defines even `__GNUC__` and does recognize #pragma GCC. Otherwise it would have to be `#ifdef __clang__ ... #elif defined(__GNUC__) ... #endif`

File: crypto/asn1/x_long.c [hunk 9]
```diff
@@ -149,6 +152,10 @@ static int long_c2i(ASN1_VALUE **pval, const unsigned char *cont, int len,
             utmp |= cont[i];
     }
     ltmp = (long)utmp;
+    if (ltmp < 0) {
+        ASN1err(ASN1_F_LONG_C2I, ASN1_R_INTEGER_TOO_LARGE_FOR_LONG);
+        return 0;
+    }
```
Comment (levitte) [inline crypto/asn1/x_long.c]: Ah, right.  I'll make sure to do that

File: test/asn1_encode_test.c [hunk 10]
```diff
@@ -0,0 +1,641 @@
...
+};
+
+typedef struct {
+    unsigned char *bytes1;
+    size_t nbytes1;
+    unsigned char *bytes2;
+    size_t nbytes2;
+} TEST_CUSTOM_DATA;
+#define CUSTOM_DATA(v)                                   \
+    { v, sizeof(v), t_one, sizeof(t_one) },       \
```
Comment (levitte) [inline test/asn1_encode_test.c]: Yes, that's misaligned.  Fixing.

File: test/asn1_encode_test.c [hunk 11]
```diff
@@ -0,0 +1,641 @@
...
+IMPLEMENT_STATIC_ASN1_ENCODE_FUNCTIONS(ASN1_LONG_DATA)
+IMPLEMENT_STATIC_ASN1_ALLOC_FUNCTIONS(ASN1_LONG_DATA)
+
+static ASN1_LONG_DATA long_expected[] = {
+    /* The following should fail on the second because it's the default */
+    { 0xff, 0, 1 }, { 0, 0, 0 }, /* t_zero */
+    { 0, 0, 0 }, { 0xff, 1, 0x7fffffff }, /* t_longundef */
+    CUSTOM_EXPECTED_SUCCESS(1, 1), /* t_one */
+    CUSTOM_EXPECTED_FAILURE,     /* t_9bytes_1 */
+#ifdef SIXTY_FOUR_BIT_LONG
```
Comment (dot-asm) [inline test/asn1_encode_test.c]: SIXTY_FOUR_BIT_LONG is just a BN configuration option, it does not necessarily mean that sizeof(long)==8. Well, it's presence does mean it, but its absence doesn't mean that sizeof(long)!=8.

File: test/asn1_encode_test.c (hunk 6 above)
Comment (levitte) [inline test/asn1_encode_test.c]: Good question.  I'll add min value.

File: test/asn1_encode_test.c (hunk 5 above)
Comment (levitte) [inline test/asn1_encode_test.c]: A remainder from earlier coding...  I'll remove it.

File: test/asn1_encode_test.c (hunk 7 above)
Comment (levitte) [inline test/asn1_encode_test.c]: I'll change it to `INT32_MAX - 1`.  Changing the value of `ASN1_LONG_UNDEF` would be baaaad, for exactly the reason you mention.  On the other hand, I would discorage any from using ASN1 LONG, for exactly the same reason.

File: test/asn1_encode_test.c (hunk 8 above)
Comment (levitte) [inline test/asn1_encode_test.c]: So basically, I can remove the clang clause entirely.  Good to know.

File: test/asn1_encode_test.c (hunk 11 above)
Comment (levitte) [inline test/asn1_encode_test.c]: No, it's not "just" a BN configuration option, it's an OpenSSL configuration option.  Also, if what you say is true, then the following snippet from `bn.h` is flaky:

``` C
/*
 * 64-bit processor with LP64 ABI
 */
# ifdef SIXTY_FOUR_BIT_LONG
#  define BN_ULONG        unsigned long
#  define BN_BYTES        8
# endif

/*
 * 64-bit processor other than LP64 ABI
 */
# ifdef SIXTY_FOUR_BIT
#  define BN_ULONG        unsigned long long
#  define BN_BYTES        8
# endif

```

Comment (dot-asm) [inline test/asn1_encode_test.c]: Yes, it's just a BN configuration option. You can compile OpenSSL on sizeof(long)==8 platform without SIXTY_FOUR_BIT_LONG and it will work. It's even supported, as we do have such platform, solaris64-sparcv9. It would be appropriate to rely on SIXTY_FOUR_BIT_LONG if BN_ULONG was used. But it isn't, long is...

Comment (dot-asm) [inline test/asn1_encode_test.c]: Just in case, yes, crypto/bn should be the only place that may rely on SIXTY_FOUR_BIT_LONG. Yes, one can argue that it should be named differently, e.g. BN_SIXTY_FOUR_BIT_LONG, but this would be out of this scope... [As would reference to SIXTY_FOUR_BIT_LONG in test/bad_dtls_test.c].

File: test/asn1_encode_test.c (hunk 7 above)
Comment (mattcaswell) [inline test/asn1_encode_test.c]: Well, that still doesn't quite do it. You want to use LONG_MAX on 64 bit and LONG_MAX -1 on 32 bit. Perhaps "# if LONG_MAX == ASN1_LONG_UNDEF" or similar.

File: test/asn1_encode_test.c (hunk 11 above)
Comment (dot-asm) [inline test/asn1_encode_test.c]: > [As would reference to SIXTY_FOUR_BIT_LONG in test/bad_dtls_test.c]

#3113.

**Comment by mattcaswell** [comment]
Travis seems to be failing for this.

**Comment by levitte** [comment]
Of course, I realise that ZFOO should really only be used with optional ASN.1 fields, or there will be errors encoding it (because the default value will not be encoded)...  this also leads me to wonder why we have the Z types to begin with (I can see why, 'cause I can't find a way to communicate optionality up and down the call stack when encoding stuff, but ...)

**Comment by levitte** [comment]
> Travis seems to be failing for this.

Because I'm using Z types as non-optional fields :-/

File: test/asn1_encode_test.c (hunk 11 above)
Comment (levitte) [inline test/asn1_encode_test.c]: Errr, so do we have a reliable way to detect in the preprocessor if long is 32- or 64-bit?  If not, I must rethink the LONG / ZLONG test data...

Comment (dot-asm) [inline test/asn1_encode_test.c]: > so do we have a reliable way to detect in the preprocessor if long is 32- or 64-bit?

Well, you can compare LONG_MAX to 32-bit maximum value, and if not equal, assume 64-bit long. Alternatively you can say `#define MAYBE_64(x) (sizeof(long)==8 ? x : 0)` and use in initializers "may-fail-or-not" initializers...

File: util/libcrypto.num [hunk 12]
```diff
@@ -4254,3 +4254,19 @@ EVP_aria_128_ctr                        4204	1_1_1	EXIST::FUNCTION:ARIA
 EVP_aria_192_ctr                        4205	1_1_1	EXIST::FUNCTION:ARIA
 UI_null                                 4206	1_1_1	EXIST::FUNCTION:UI
 EC_KEY_get0_engine                      4207	1_1_1	EXIST::FUNCTION:EC
+INT32_it                                4208	1_1_1	EXIST:!EXPORT_VAR_AS_FUNCTION:VARIABLE:
```
Comment (dot-asm) [inline util/libcrypto.num]: Request is marked as 1.1.0 and master. How is this supposed to fly in 1.1.0?

Comment (levitte) [inline util/libcrypto.num]: Not at all, I forgot to adjust this file, 's all.

**Comment by levitte** [comment]
Hopefully, I've addressed all concerns so far

**Comment by levitte** [comment]
Passes all checks. Have another look?

File: test/asn1_encode_test.c (hunk 11 above)
Comment (levitte) [inline test/asn1_encode_test.c]: Side note: speaking of this, it seems we need to reword `Configurations/README`, as it currently says this:

```
                           SIXTY_FOUR_BIT       processor registers
                                                are 64 bits, long is
                                                32 bits, long long is
                                                64 bits.
                           SIXTY_FOUR_BIT_LONG  processor registers
                                                are 64 bits, long is
                                                64 bits.
```

Comment (levitte) [inline test/asn1_encode_test.c]: > Well, you can compare LONG_MAX to 32-bit maximum value, and if not equal, assume 64-bit long.

Tried `#if LONG_MAX > INT32_MAX`, and it works...  but humm, I have some concerns, 'cause `sizeof()` isn't a preprocessor operator (right?), so *if* our definition of `INT32_MAX` is used here, the build should fail.  No?

Comment (dot-asm) [inline test/asn1_encode_test.c]: Suggestion was to check for equality as in `!=` ;-) As for sizeof(), it wasn't about pre-processor in #if-#else sense, but using conditional expressions in initializes. I.e. it was meant as *alternative* to `#if LONG_MAX != INT32_MAX`

Comment (mattcaswell) [inline test/asn1_encode_test.c]: >  if our definition of INT32_MAX is used here, the build should fail.

Actually that suggests to me that our definitions are wrong - they shouldn't use sizeof(). It would be quite simple to fix that because we know the sizes of these defined size types (with the exception of SIZE_MAX which is more tricky).

File: test/asn1_encode_test.c [hunk 13]
```diff
@@ -0,0 +1,681 @@
...
+    CUSTOM_EXPECTED_FAILURE,     /* t_5bytes_1 */
+    CUSTOM_EXPECTED_FAILURE,     /* t_4bytes_1 (too large positive) */
+#endif
+    CUSTOM_EXPECTED_SUCCESS(INT32_MAX - 1, INT32_MAX -1), /* t_4bytes_2 */
+    CUSTOM_EXPECTED_FAILURE,     /* t_4bytes_3_pad (illegal padding) */
+    CUSTOM_EXPECTED_SUCCESS(INT32_MIN, INT32_MIN), /* t_4bytes_4_neg */
+    CUSTOM_EXPECTED_FAILURE,     /* t_4bytes_5_negpad (illegal padding) */
+};
+static ASN1_LONG_DATA long_encdec_data[] = {
+    ENCDEC_ARRAY(INT32_MAX - 1, INT32_MAX, INT32_MIN, INT32_MIN),
```
Comment (mattcaswell) [inline test/asn1_encode_test.c]: This still doesn't look right to me. These should be using LONG_MAX and LONG_MIN *unless* LONG_MAX == INT32_MAX in which case max is LONG_MAX - 1.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Looks good. I just have one comment below, and there's also the outstanding issue of the "ifdef SIXTY_FOUR_BIT_LONG" line which needs resolution

File: test/asn1_encode_test.c (hunk 11 above)
Comment (levitte) [inline test/asn1_encode_test.c]: @dot-asm, conditional initializers will be problematic with values like `0x1ffffffff` on 32-bit machines.  No can do (at least with some semblance of elegance).

Comment (levitte) [inline test/asn1_encode_test.c]: I'm starting to think that we should generate numbers.h with the help of a very simple C program that provides the sizes.  That will fix the problem with things like SIZE_MAX.

Comment (dot-asm) [inline test/asn1_encode_test.c]: > conditional initializers will be problematic with values like 0x1ffffffff on 32-bit machines.

Not if you use explicit cast, `(long)0x1ffffffff` :-)

**Comment by levitte** [comment]
Ok, I finally got rid of SIXTY_FOUR_BIT_LONG.  I hope the builds will agree with the result.

File: test/asn1_encode_test.c (hunk 13 above)
Comment (levitte) [inline test/asn1_encode_test.c]: Fixed

**Comment by levitte** [comment]
No complaints from Travis, that's good :-)

**Comment by levitte** [comment]
And no complaint from Appveyor.  Just to be safe, I'm doing a VMS build too.

**Comment by levitte** [comment]
Builds and tests fine on VMS too.

Last review, folks?

**Review by mattcaswell — APPROVED** [review]
One last thought. In theory this should be back portable without any impact. I'd kind of like the reassurance that a 1.1.0 session file created before this change can be loaded successfully with a 1.1.0 version with this change. Could you try it out? Assuming you do (and it works!) - approved.

**Comment by levitte** [comment]
I should make a separate PR for 1.1.0...  I just realised that `asn1_encode_test.c` can't be backported to 1.1.0, 'cause we're lacking the new test programming framework there...

However, I can confirm that `test/recipes/25-test_sid.t` runs flawlessly on 1.1.0 with the patches that don't involve `test/asn1_encode_test`.  Are there specific tests other than that that you would like me to try?

**Comment by mattcaswell** [comment]
> However, I can confirm that test/recipes/25-test_sid.t runs flawlessly on 1.1.0 with the patches that don't involve test/asn1_encode_test. Are there specific tests other than that that you would like me to try?

Ahh. Right. I see that is using a session file that is already generated - so that should be fine. Perhaps additionally create a session file with this patch, and check it can be read by 1.1.0 prior to this patch?

**Comment by levitte** [comment]
```
: ; ./util/shlib_wrap.sh apps/openssl version
OpenSSL 1.1.0f-dev  xx XXX xxxx
: ; ./util/shlib_wrap.sh apps/openssl s_client -sess_out foo.sid -connect www.openssl.org:443 < /dev/null > /dev/null
depth=1 C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
verify error:num=20:unable to get local issuer certificate
DONE
: ; ls -l foo.sid
-rw-r--r-- 1 levitte levitte 2221 apr  4 11:22 foo.sid
: ; type openssl
openssl är /usr/bin/openssl
: ; openssl version
OpenSSL 1.1.0e  16 Feb 2017
: ; openssl sess_id -in foo.sid -text -noout
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : C030
    Session-ID: 530C7D2C6DDE96932E3E443B02A11EA53BBBAE3453E3AD8FED6E1439CE8EB300
    Session-ID-ctx: 
    Master-Key: DA4BFF7E81AED7D287A117F0D27AD4959AD65F487C3CD4E55E4F6C6D7BF7E6241E939F6427EE45F80652E41B441767C4
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    0000 - 00 00 0c 1e 6d b0 c5 ba-43 7e 29 08 f6 a9 20 3d   ....m...C~)... =
    0010 - f6 e7 5c 87 dc 52 af 4a-3e 53 2c 81 0d c4 42 0e   ..\..R.J>S,...B.
    0020 - e7 b6 a0 29 f7 a1 bf ad-00 a3 73 db b1 53 7b 3c   ...)......s..S{<
    0030 - a6 5f e4 4d 12 08 34 3d-2a 2f 5e a8 69 42 b5 1b   ._.M..4=*/^.iB..
    0040 - c4 35 d0 9e d0 7f 0b 45-f0 7f db d3 6d 67 5c 6b   .5.....E....mg\k
    0050 - c9 06 97 e5 c0 6e 45 19-b7 79 a3 a1 77 ed 6e 53   .....nE..y..w.nS
    0060 - bf 34 09 d5 7d 3a a3 60-79 32 69 7f a2 64 1b 67   .4..}:.`y2i..d.g
    0070 - 17 bc 6c 95 a5 22 6d 5e-d9 5b a6 b9 83 a8 8e 2e   ..l.."m^.[......
    0080 - 9d 34 a2 0f 08 4f 2e 8c-82 19 c4 b2 64 e7 15 44   .4...O......d..D
    0090 - 56 1e dd 2d 42 34 1b 95-78 5d ec 4c 4f 90 63 f1   V..-B4..x].LO.c.

    Start Time: 1491297762
    Timeout   : 7200 (sec)
    Verify return code: 20 (unable to get local issuer certificate)
    Extended master secret: no
```

Looks good enough?

**Comment by mattcaswell** [comment]
LGTM. +1

**Comment by levitte** [comment]
Good.  Because of the differences, I've removed the 1.1.0 label here and created #3120.  Please give it a quick review as well.

**Comment by levitte** [comment]
Merged to master.  93f7d6fc10b75814d90d804edb56947cacf8964e to 79b3452faf04f2572f57eb37b618cc603d9983da
